metastore/src/java/org/apache/hadoop/hive/metastore/ObjectStore.java

```diff
@@ -8268,29 +8279,90 @@ public NotificationEventResponse getNextNotification(NotificationEventRequest rq
     }
   }
 
+  private void lockForUpdate() throws MetaException {
+    String selectQuery = "select NEXT_EVENT_ID from NOTIFICATION_SEQUENCE";
+    String selectForUpdateQuery = sqlGenerator.addForUpdateClause(selectQuery);
+    new RetryingExecutor(hiveConf, () -> {
+      Query query = pm.newQuery("javax.jdo.query.SQL", selectForUpdateQuery);
+      query.setUnique(true);
+      // only need to execute it to get db Lock
+      query.execute();
+    }).run();
+  }
+
+  static class RetryingExecutor {
+    interface Command {
+      void process() throws Exception;
+    }
+
+    private static Logger LOG = LoggerFactory.getLogger(RetryingExecutor.class);
+    private final int maxRetries;
+    private final long sleepInterval;
+    private int currentRetries = 0;
+    private final Command command;
+
+    RetryingExecutor(Configuration config, Command command) {
+      this.maxRetries = config.getInt(ConfVars.NOTIFICATION_SEQUENCE_LOCK_MAX_RETRIES.name(),
+          ConfVars.NOTIFICATION_SEQUENCE_LOCK_MAX_RETRIES.defaultIntVal);
+      this.sleepInterval = config.getTimeDuration(
+          ConfVars.NOTIFICATION_SEQUENCE_LOCK_RETRY_SLEEP_INTERVAL.name(),
+          ConfVars.NOTIFICATION_SEQUENCE_LOCK_RETRY_SLEEP_INTERVAL.defaultLongVal,
+          TimeUnit.MILLISECONDS
+      );
+      this.command = command;
+    }
+
+    public void run() throws MetaException {
+      while (true) {
+        try {
+          command.process();
+          break;
+        } catch (Exception e) {
+          LOG.info("retried / max : {} / {}", currentRetries, maxRetries);
+          if (currentRetries >= maxRetries) {
+            LOG.error("failed to execute Command :: ", e);
```

Comment:
We also need a better error message so that users can understand what happened and what to do in case they see this error.

Can I suggest this message?
- Couldn't acquire the DB log notification lock because we reached the maximum # of retries: {} retries. If this happens too often, then is recommended to increase the maximum number of retries on the hive.notification.sequence.lock.max.retries configuration.

Reply:
done